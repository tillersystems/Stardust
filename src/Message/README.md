# Message

### Usage

```jsx
import { Message } from '@tillersystems/stardust';
```

<!-- STORY -->

### Properties

- `className` - className needed by styled components.
- `description` - An alert can have a message description.
- `onClose` -An alert can have a clickable button to close it.
- `type` - The type of the message box.

| propName    |                       propType                       | defaultValue | isRequired |
| ----------- | :--------------------------------------------------: | :----------: | :--------: |
| `className` |                       `string`                       |     `''`     |            |
| `message`   |                       `string`                       |              |     \*     |
| `onClose`   |                        `func`                        |  `() => {}`  |            |
| `type`      | `enum: 'info', 'success', 'warning', 'error' 'info'` |    `info`    |            |

### Example

```jsx
import { Message } from '@tillersystems/stardust';

render() {
  return (
    <Message description="this is a message" type="success">
  );
}
```
