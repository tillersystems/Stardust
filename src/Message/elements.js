import styled from 'styled-components';

export const Description = styled.span`
  flex: 1;
  text-align: left;
  margin-left: 2rem;
`;
