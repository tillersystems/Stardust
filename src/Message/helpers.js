export const getStatusBackgroundColor = ({ palette }, type) => {
  return {
    info: palette.primary.default,
    success: palette.success.default,
    error: palette.failure.default,
    warning: palette.warning.default,
  }[type];
};
