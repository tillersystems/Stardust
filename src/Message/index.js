import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import { Icon, Theme } from '..';
import { Description } from './elements';
import { STATUS_ICON_NAMES } from './constants';
import { getStatusBackgroundColor } from './helpers';

/**
 * Message
 *
 * This component is in charge of displaying
 * an Message for a user
 *
 * @param {bool} closable // Whether it is possible to close the Message box.
 * @param {string} description // An alert can have a message description.
 * @param {func} onClose // An message can have a clickable button to close it.
 * @param {enum} type // The type of the message box.
 *
 * @return {jsx}
 */

const Message = ({ className, description, onClose, type }) => (
  <div className={className} type={type}>
    <Icon name={STATUS_ICON_NAMES[type]} color={Theme.palette.white} />
    <Description>{description}</Description>
    {onClose && (
      <Icon
        tabIndex="0"
        role="button"
        name="cross"
        color={Theme.palette.white}
        size="18px"
        onClick={onClose}
        data-testid="close-button"
      />
    )}
  </div>
);
/**
 * PropTypes Validation
 */
const { func, oneOf, string } = PropTypes;
Message.propTypes = {
  className: string,
  description: string.isRequired,
  onClose: func,
  type: oneOf(['success', 'info', 'warning', 'error']),
};

/**
 * Default props
 */
Message.defaultProps = {
  className: '',
  onClose: null,
  type: 'info',
};

const styledMessage = styled(Message)`
  display: flex;
  align-items: center;

  padding: 1.3rem;

  border-radius: 0.5rem;

  color: ${({ theme: { palette } }) => palette.white};

  cursor: pointer;

  ${({ theme, type }) =>
    css`
      background: ${getStatusBackgroundColor(theme, type)};
    `};
`;

styledMessage.displayName = 'Message';

export default styledMessage;
