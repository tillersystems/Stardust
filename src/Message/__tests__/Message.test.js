import React from 'react';
import { fireEvent } from 'react-testing-library';

import Message from '..';

describe('<Message />', () => {
  test('should render without a problem', () => {
    const { container } = render(<Message description="this is a message" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render a success alert', () => {
    const { container } = render(<Message description="this is a message" type="success" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render a info alert', () => {
    const { container } = render(<Message description="this is a message" type="info" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render a warning alert', () => {
    const { container } = render(<Message description="this is a message" type="warning" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render a error alert', () => {
    const { container } = render(<Message description="this is a message" type="error" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should call onClose handler', () => {
    const handleOnClose = jest.fn();
    const closeText = 'close';
    const { getByText } = render(
      <Message
        description="this is a message"
        type="error"
        onClose={handleOnClose}
        onCloseText={closeText}
      />,
    );
    const dismissMessageNode = getByText(closeText);

    fireEvent.click(dismissMessageNode);

    expect(handleOnClose).toHaveBeenCalled();
  });
});
