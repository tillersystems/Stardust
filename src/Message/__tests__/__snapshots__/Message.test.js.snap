// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Message /> should render a error alert 1`] = `
.c1 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  text-align: left;
  margin-left: 2rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1.3rem;
  border-radius: 0.5rem;
  color: hsl(0,100%,100%);
  cursor: pointer;
  background: hsl(6,79%,65%);
}

.c0 button {
  font-size: 1.4rem;
  background: none;
  color: hsl(0,100%,100%);
  cursor: pointer;
}

<div
  class="c0"
  data-testid="message"
  type="error"
>
  <span
    aria-label="this is a message"
    role="alert"
  />
  <div
    aria-hidden="true"
    class="c1"
    data-testid="iconContainer"
    height="20"
    width="20"
  >
    <svg
      data-testid="iconSvg"
      height="20"
      style="display: block;"
      viewBox="0 0 512 512"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 492C125.66 492 20 386.34 20 256S125.66 20 256 20s236 105.66 236 236-105.66 236-236 236zm0-138.373c10.784 0 19.525-8.742 19.525-19.525 0-10.784-8.741-19.526-19.525-19.526-10.784 0-19.525 8.742-19.525 19.526 0 10.783 8.741 19.525 19.525 19.525zm-19.525-175.729V256c0 10.784 8.741 19.525 19.525 19.525 10.784 0 19.525-8.741 19.525-19.525v-78.102c0-10.783-8.741-19.525-19.525-19.525-10.784 0-19.525 8.742-19.525 19.525z"
        data-testid="iconSvgPath"
        fill="hsl(0,100%,100%)"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <span
    class="c2"
  >
    this is a message
  </span>
</div>
`;

exports[`<Message /> should render a info alert 1`] = `
.c1 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  text-align: left;
  margin-left: 2rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1.3rem;
  border-radius: 0.5rem;
  color: hsl(0,100%,100%);
  cursor: pointer;
  background: hsl(200,74%,46%);
}

.c0 button {
  font-size: 1.4rem;
  background: none;
  color: hsl(0,100%,100%);
  cursor: pointer;
}

<div
  class="c0"
  data-testid="message"
  type="info"
>
  <span
    aria-label="this is a message"
    role="alert"
  />
  <div
    aria-hidden="true"
    class="c1"
    data-testid="iconContainer"
    height="20"
    width="20"
  >
    <svg
      data-testid="iconSvg"
      height="20"
      style="display: block;"
      viewBox="0 0 512 512"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 492C125.66 492 20 386.34 20 256S125.66 20 256 20s236 105.66 236 236-105.66 236-236 236zm21.333-150.667V256c0-11.782-9.55-21.333-21.333-21.333-11.782 0-21.333 9.55-21.333 21.333v85.333c0 11.782 9.55 21.334 21.333 21.334 11.782 0 21.333-9.552 21.333-21.334zM256 192c11.782 0 21.333-9.551 21.333-21.333 0-11.782-9.55-21.334-21.333-21.334-11.782 0-21.333 9.552-21.333 21.334S244.217 192 256 192z"
        data-testid="iconSvgPath"
        fill="hsl(0,100%,100%)"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <span
    class="c2"
  >
    this is a message
  </span>
</div>
`;

exports[`<Message /> should render a success alert 1`] = `
.c1 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  text-align: left;
  margin-left: 2rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1.3rem;
  border-radius: 0.5rem;
  color: hsl(0,100%,100%);
  cursor: pointer;
  background: hsl(89,53%,52%);
}

.c0 button {
  font-size: 1.4rem;
  background: none;
  color: hsl(0,100%,100%);
  cursor: pointer;
}

<div
  class="c0"
  data-testid="message"
  type="success"
>
  <span
    aria-label="this is a message"
    role="alert"
  />
  <div
    aria-hidden="true"
    class="c1"
    data-testid="iconContainer"
    height="20"
    width="20"
  >
    <svg
      data-testid="iconSvg"
      height="20"
      style="display: block;"
      viewBox="0 0 512 512"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M411.582 134.248c8.33-8.33 21.838-8.33 30.17 0 8.33 8.332 8.33 21.84 0 30.17L207.085 399.085c-8.331 8.331-21.839 8.331-30.17 0L70.248 292.418c-8.33-8.33-8.33-21.838 0-30.17 8.332-8.33 21.84-8.33 30.17 0L192 353.83l219.582-219.582z"
        data-testid="iconSvgPath"
        fill="hsl(0,100%,100%)"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <span
    class="c2"
  >
    this is a message
  </span>
</div>
`;

exports[`<Message /> should render a warning alert 1`] = `
.c1 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  text-align: left;
  margin-left: 2rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1.3rem;
  border-radius: 0.5rem;
  color: hsl(0,100%,100%);
  cursor: pointer;
  background: hsl(34,91%,65%);
}

.c0 button {
  font-size: 1.4rem;
  background: none;
  color: hsl(0,100%,100%);
  cursor: pointer;
}

<div
  class="c0"
  data-testid="message"
  type="warning"
>
  <span
    aria-label="this is a message"
    role="alert"
  />
  <div
    aria-hidden="true"
    class="c1"
    data-testid="iconContainer"
    height="20"
    width="20"
  >
    <svg
      data-testid="iconSvg"
      height="20"
      style="display: block;"
      viewBox="0 0 512 512"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 492C125.66 492 20 386.34 20 256S125.66 20 256 20s236 105.66 236 236-105.66 236-236 236zm0-138.373c10.784 0 19.525-8.742 19.525-19.525 0-10.784-8.741-19.526-19.525-19.526-10.784 0-19.525 8.742-19.525 19.526 0 10.783 8.741 19.525 19.525 19.525zm-19.525-175.729V256c0 10.784 8.741 19.525 19.525 19.525 10.784 0 19.525-8.741 19.525-19.525v-78.102c0-10.783-8.741-19.525-19.525-19.525-10.784 0-19.525 8.742-19.525 19.525z"
        data-testid="iconSvgPath"
        fill="hsl(0,100%,100%)"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <span
    class="c2"
  >
    this is a message
  </span>
</div>
`;

exports[`<Message /> should render without a problem 1`] = `
.c1 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  text-align: left;
  margin-left: 2rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1.3rem;
  border-radius: 0.5rem;
  color: hsl(0,100%,100%);
  cursor: pointer;
}

.c0 button {
  font-size: 1.4rem;
  background: none;
  color: hsl(0,100%,100%);
  cursor: pointer;
}

<div
  class="c0"
  data-testid="message"
  type="info"
>
  <span
    aria-label="this is a message"
    role="alert"
  />
  <div
    aria-hidden="true"
    class="c1"
    data-testid="iconContainer"
    height="20"
    width="20"
  >
    <svg
      data-testid="iconSvg"
      height="20"
      style="display: block;"
      viewBox="0 0 512 512"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 492C125.66 492 20 386.34 20 256S125.66 20 256 20s236 105.66 236 236-105.66 236-236 236zm21.333-150.667V256c0-11.782-9.55-21.333-21.333-21.333-11.782 0-21.333 9.55-21.333 21.333v85.333c0 11.782 9.55 21.334 21.333 21.334 11.782 0 21.333-9.552 21.333-21.334zM256 192c11.782 0 21.333-9.551 21.333-21.333 0-11.782-9.55-21.334-21.333-21.334-11.782 0-21.333 9.552-21.333 21.334S244.217 192 256 192z"
        data-testid="iconSvgPath"
        fill="hsl(0,100%,100%)"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <span
    class="c2"
  >
    this is a message
  </span>
</div>
`;
